catalogd/internal/storage/index.go

```diff
@@ -0,0 +1,133 @@
+package storage
+
+import (
+	"cmp"
+	"encoding/json"
+	"fmt"
+	"io"
+	"slices"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/operator-framework/operator-registry/alpha/declcfg"
+)
+
+type index struct {
+	BySchema  map[string][]section `json:"by_schema"`
+	ByPackage map[string][]section `json:"by_package"`
+	ByName    map[string][]section `json:"by_name"`
+}
+
+type section struct {
+	offset int64
+	length int64
+}
+
+func (s *section) MarshalJSON() ([]byte, error) {
+	return []byte(fmt.Sprintf(`[%d,%d]`, s.offset, s.length)), nil
+}
+
+func (s *section) UnmarshalJSON(b []byte) error {
+	vals := [2]int64{}
+	if err := json.Unmarshal(b, &vals); err != nil {
+		return err
+	}
+	s.offset = vals[0]
+	s.length = vals[1]
+	return nil
+}
+
+func (i index) Size() int64 {
+	size := 0
+	for k, v := range i.BySchema {
+		size += len(k) + len(v)*16
+	}
+	for k, v := range i.ByPackage {
+		size += len(k) + len(v)*16
+	}
+	for k, v := range i.ByName {
+		size += len(k) + len(v)*16
+	}
+	return int64(size)
+}
+
+func (i index) Get(r io.ReaderAt, schema, packageName, name string) (io.ReadSeeker, bool) {
+	sectionSet := i.getSectionSet(schema, packageName, name)
+
+	sections := sectionSet.UnsortedList()
+	slices.SortFunc(sections, func(a, b section) int {
+		return cmp.Compare(a.offset, b.offset)
+	})
+
+	srs := make([]io.ReadSeeker, 0, len(sections))
+	for _, s := range sections {
+		sr := io.NewSectionReader(r, s.offset, s.length)
+		srs = append(srs, sr)
+	}
+	return newMultiReadSeeker(srs...), true
+}
+
+func (i *index) getSectionSet(schema, packageName, name string) sets.Set[section] {
```

Comment:
One thing that occurs to me is that the indexed sections are always guaranteed to be sorted by the offset. We could likely take advantage of that fact and implement a more performant algorithm that can step directly through each slice rather than building a set for each slice, performing a set intersection, the converting back into a slice and resorting.

But this is something we can capture as a follow-up.